apps/api/test/04-report-4.e2e-spec.ts

```diff
@@ -0,0 +1,265 @@
+  // Tracks applicants id's and to later delete status/milestones
+  const applicantIds: string[] = [];
+  const jobIds: string[] = [];
+  let app: INestApplication;
+
+  beforeAll(async () => {
+    const moduleFixture: TestingModule = await Test.createTestingModule({
+      imports: [AppModule],
+    })
+      .overrideGuard(AuthGuard)
+      .useValue({ canActivate })
+      .compile();
+
+    app = moduleFixture.createNestApplication();
+    await app.init();
+
+    setApp(app);
+  });
+
+  afterAll(async () => {
+    await app.close();
+  });
+
+  it('Status 200 on report 4 response', async () => {
+    const { status } = await request(app.getHttpServer()).get(URLS.REPORT4);
+    expect(status).toBe(200);
+  });
+
+  it('Add applicants and give each a different status/milestone', async () => {
+    const statusIds: string[] = [];
+
+    const { body: before } = await request(app.getHttpServer()).get(URLS.REPORT4);
+
+    // What the report is expected to look like. Used for comparison at the end of the test
+    const expectedResult = before.map((item: { applicants: string; status: string }) => {
+      const stat = item.status;
+      let result = parseInt(item.applicants) + 1;
+      // Accounts for other two BCCNM Licenses
+      if (stat === 'Granted provisional licensure' || stat === 'Granted full licensure') {
+        result++;
+      }
+      // Withdraw status should not be incremented in report ouput
+      else if (stat === STATUS.WITHDREW_FROM_PROGRAM) {
+        result--;
+      }
+      // NCAS count increases with COMPLETED_CBA and COMPLETED_SLA
+      // NNAS count increases with RECEIVED_NNAS_REPORT and SUBMITTED_DOCUMENTS
+      else if (stat === STATUS.REFERRED_TO_NCAS || stat === STATUS.APPLIED_TO_NNAS) {
+        result += 2;
+      }
+      // Referred to Additional Education makes this report count increase
+      else if (stat === STATUS.COMPLETED_ADDITIONAL_EDUCATION) {
+        result++;
+      }
+      return {
+        ...item,
+        applicants: result.toString(),
+      };
+    });
+
+    // Add applicants and give milestones/status
+    for (const status of LIC_REG_STAGE) {
+      // Job id only needed for WITHDREW_FROM_PROGRAM
+      let jobId = '';
+      const applicant = await addApplicant(getApplicant());
+      applicantIds.push(applicant.id);
+
+      if (status === STATUS.WITHDREW_FROM_PROGRAM) {
+        const { id } = await addJob(applicant.id, {
+          ha_pcn: haPcnArray[0],
+          job_id: '148593',
+          recruiter_name: 'Test 4',
+        });
+        jobId = id;
+      }
+      const milestone = await addMilestone(applicant.id, jobId, await getStatus(status));
+      statusIds.push(milestone.id);
+    }
+
+    const { body: after } = await request(app.getHttpServer()).get(URLS.REPORT4);
+    expect(after).toStrictEqual(expectedResult);
+
+    // Resetting milestones to original
+    for (let i = 0; i <= LIC_REG_STAGE.length - 1; i++) {
```

Comment:
This test may not be affected by other tests. If other tests fail because of applicants and milestones added by this test, why don't we fix them instead of cleaning test data?

Reply:
Just to clarify, you mean if other tests fail because of my tests, then I should fix the other tests and not clean the test data in my test? I figured that isolating my test would be best so others don't have to worry about fixing my test in the future. Do you think this will cause issues in the future?

Reply:
By other tests I mean tests other than report 4 test

Reply:
I meant that we just need to write tests so that they will not be affected by other tests or will not affect others, but not both. In my opinion, the former is harder than the latter.